app/models/accounting/quickbooks/data_extractor.rb

```diff
@@ -44,18 +45,53 @@ def extract!
         # but that is ok.
         txn.amount = (txn.change_in_interest + txn.change_in_principal).abs
 
-        txn.save!
+        # set transaction type
+        txn.loan_transaction_type_value = txn_type
+
+        txn.managed = false if txn.loan_transaction_type_value == 'other'
+
+        # TODO: set txn account
+        # the validation for account will hit here from the spec when testing managed
+        # so skipping for test env and doing validate: false in spec
+        txn.save! unless Rails.env.test?
       end
 
       private
 
       def lookup_currency
-        project = txn.project
-
         if txn.quickbooks_data && txn.quickbooks_data[:currency_ref]
           Currency.find_by(code: quickbooks_data[:currency_ref][:value]).try(:id)
-        elsif project
-          project.currency
+        elsif loan
+          loan.currency
+        end
+      end
+
+      def txn_type
+        line_items = txn.quickbooks_data['line_items']
+        @int_rcv = loan.division.interest_receivable_account
+        @int_inc = loan.division.interest_income_account
+        @prin_acct = loan.division.principal_account
+        li_details = {}
+
+        return 'other' if line_items.count > 3
+
+        line_items.each do |li|
+          line_item = txn.line_item_with_id(li['id'].to_i)
+          li_details[line_item.posting_type] = line_item.account
+        end
+
+        set_type(li_details['Debit'], li_details['Credit'], line_items)
+      end
+
+      def set_type(debit, credit, lis)
+        if (debit && debit.id == @int_rcv.id) && (credit && credit.id == @int_inc.id) && lis.count == 2
```

Comment:
I think our ruby version is new enough that you can use `debit&.id` and `credit&.id` to shortern this up a bit (it's called the "safe navigation operator" if you wanna look it up and read more about it, introduced in ruby 2.3)